stan/math/rev/fun/columns_dot_product.hpp

```diff
@@ -19,12 +20,43 @@ template <typename Mat1, typename Mat2,
           require_any_eigen_vt<is_var, Mat1, Mat2>* = nullptr>
 inline Eigen::Matrix<return_type_t<Mat1, Mat2>, 1, Mat1::ColsAtCompileTime>
 columns_dot_product(const Mat1& v1, const Mat2& v2) {
-  check_matching_sizes("dot_product", "v1", v1, "v2", v2);
-  Eigen::Matrix<var, 1, Mat1::ColsAtCompileTime> ret(1, v1.cols());
-  for (size_type j = 0; j < v1.cols(); ++j) {
-    ret.coeffRef(j) = dot_product(v1.col(j), v2.col(j));
+  check_matching_sizes("columns_dot_product", "v1", v1, "v2", v2);
+
+  ref_type_t<Mat1> v1_ref = v1;
+  ref_type_t<Mat2> v2_ref = v2;
+
+  arena_matrix<promote_scalar_t<double, Mat1>> arena_v1_val = value_of(v1_ref);
+  arena_matrix<promote_scalar_t<double, Mat2>> arena_v2_val = value_of(v2_ref);
+
+  arena_matrix<promote_scalar_t<var, Mat1>> arena_v1;
+  arena_matrix<promote_scalar_t<var, Mat2>> arena_v2;
+
+  if (!is_constant<Mat1>::value) {
+    arena_v1 = v1_ref;
+  }
+
+  if (!is_constant<Mat2>::value) {
+    arena_v2 = v2_ref;
   }
-  return ret;
+
+  Eigen::RowVectorXd out_val(arena_v1_val.cols());
+  for (size_t m = 0; m < arena_v1_val.cols(); ++m)
+    out_val.coeffRef(m) = arena_v1_val.col(m).dot(arena_v2_val.col(m));
```

Comment:
See the other comment that talks about the row version of this for the splainer'
```suggestion
Eigen::VectorXd out_val = (arena_v1_val.cwiseProduct(arena_v2_val)).colwise().sum();
```